### ISSUE TITLE
More flexible API for record data deserialization

### ISSUE THREAD

**Issue by not-my-profile** [NLnetLabs/domain#105]
I would love to use this library for a small educational name server I am working on, however the API does not quite seem to cover my use case (yet? :). I want to provide a command-line interface to add records while the server is running, e.g:

```
myclient example.com add mx 10 alt1.example.com.
```

The only provided API to parse `MasterRecordData` from a human friendly string seems to be [`MasterRecordData::scan()`](https://docs.rs/domain/0.6.1/domain/rdata/enum.MasterRecordData.html), powered by the [`Scan`](https://docs.rs/domain/0.6.1/domain/master/scan/trait.Scan.html) trait of the experimental `master` module. While attempting to use this API, I stumbled upon several issues:

* unhelpful error messages
  * attempting to parse `x` as MX data yields `unexpected 'x'`, it would be nice to tell the user what was expected (numeric MX priority)
  * attempting to parse `99999` as MX data yields `illegal integer`, it would be nice to tell the user why the integer is illegal (priority may not be greater than 65535)
  * `unespected end of file` doesn't tell you what was expected
* The `Scan` trait isn't suited to parse command-line arguments since Scanner assumes the context of a Zone file.
   * attempting to parse `10 foo` as MX data yields `owner @ without preceding $ORIGIN` (which I'd argue is confusing even in a zone file but is of course completely out of place for command-line arguments)
* I would like users to be able to directly provide binary strings without needing to escape them, that's not possible with Scanner.
   * For example `TXT` records can contain arbitrary binary data, I want users to be able to directly provide an arbitrary TXT value as a command-line argument (which don't need to be UTF-8).

So I think it would be great if this library could somehow support both parsing records from zone files as well as parsing them from e.g. an `Iterator<Item=OsString>` (where each item represents one part of the record data).

I guess this could be achieved by changing the `Scan` trait, that record data types implement, to take a new `Scanner` trait (instead of a zone-specific scanner). I am not quite sure yet how this `Scanner` trait would look like exactly but I think it would need to abstract over the zone-file specifics like `$ORIGIN`, by e.g. having a `scan_absolute_name(&mut self) -> Result<Dname, ScanError>` method.

Do you think this is worth pursuing?

**Comment by partim** [comment]
I am not really happy with the current zonefile parser either. It is rather convoluted and, worse, slow. Plus, it only works with `Bytes` right now which isn’t ideal in all cases. We have a plan to entirely re-write it, possibly on top of something like _nom,_ but sadly at this very moment we have no time to do it.

That said, I think the current scanner should work for your purpose. It doesn’t actually keep state, so you could create a new one for each item in your iterator, provided these represent single ’lines’. All you need to do to circumvent the owner-without-origin error is to set an origin via `Scanner::set_origin`. You probably want the root domain in your example, so `Dname::root()` should do. The reason you get the error is that your are providing a relative domain name (the `foo` doesn’t end with a dot), so the parser needs a base name to append to it.

I am a bit torn on the idea to provide binary strings directly. I guess you want to avoid the confusion of having to use backslashes in a shell? This use case needs a bit of thinking, though. The reason the current scanner works on top of `char` and not `u8` is to do with the $include directive and translating its value into a file system path. But I think it should be possible to rethink the `CharSource` trait to be able to deal with this somehow – it just never occurred to me that someone may want to be able to inject arbitrary binary data.

Long story short: I absolutely would want to improve the zonefile parser and make it handle additional use cases. What I don’t quite know is the timeline for that.

**Comment by not-my-profile** [comment]
I guess my point is that for my application I don't need to parse zone files at all, so I don't really want to work around a zone file parser. Also note that I wouldn't want to automatically convert relative to absolute domain names, I just want a descriptive error message, e.g. `absolute domain name required`. I could of course map `SyntaxError::NoOrigin` but then I am again working around something I don't actually need.

>I guess you want to avoid the confusion of having to use backslashes in a shell?

Shells already provide their own character escaping mechanism e.g. `echo $'\x1f'`. I would like users to be able to use this universal syntax instead of having to learn a tool-specific escape mechanism. Taking raw binary strings of course also increases the interoperability with other programs.

Are you opposed to the idea of introducing an intermediary trait to allow record data types to be built from binary strings without any zone file parser? If not I could attempt to introduce this abstraction.

**Comment by partim** [comment]
> Are you opposed to the idea of introducing an intermediary trait to allow record data types to be built from binary strings without any zone file parser?

I’d love to have a more flexible API for deserializing records. Could you describe your idea a bit more? Ie., what do you mean by ’raw binary strings’? Providing something like MX’s priority that way seems a bit tedious.

**Comment by not-my-profile** [comment]
The abstraction I have in mind is something like:

```rs
trait Deserializer<N> {
    type Error;
    
    fn deserialize_u16(&mut self) -> Result<u16, Self::Error>;
    fn deserialize_domain_name(&mut self) -> Result<N, Self::Error>;
    // etc. whatever RDATA formats need
}

trait DeserializeRdata<N> {
    fn deserialize<D: Deserializer<N>>(d: &mut D) -> Result<Self, D::Error> where Self: Sized;
}

struct Mx<N> {
    preference: u16,
    exchange: N,
}

impl<N> DeserializeRdata<N> for Mx<N> {
    fn deserialize<D: Deserializer<N>>(d: &mut D) -> Result<Self, D::Error> {
        Ok(Self{preference: d.deserialize_u16()?, exchange: d.deserialize_domain_name()?})
    }
}
```

**Comment by partim** [comment]
But of course! We should probably use different terms to distinguish this from serde. `DeserializeRdata` can probably just be `Deserialize` (or whatever name we choose) and be used for records and record sets, too.

I am not at all opposed to reusing the `Scan` and `Scanner` names here and changing everything accordingly. This is much better than the current approach.

**Comment by not-my-profile** [comment]
I'm glad you like it. Good call on using other terminology than serde. I am also alright with sticking with Scan(ner). I don't quite see the appeal of generalizing this to records and record sets, since they have a fixed format that doesn't require such flexibility.

**Comment by partim** [comment]
They sort of do indirectly through the record data. The ‘header’ of a record (i.e., domain name, TTL, class, record type) also needs to parse tokens, so using the same API makes sense.

**Comment by not-my-profile** [comment]
I think it's good practice to keep traits minimal for a specific purpose. As far as I know RDATA never contains record types or DNS CLASSes, so I don't think these should be provided by the trait used for RDATA deserialization.

**Comment by partim** [comment]
But these are just string tokens (or ‘words’ in the current scanner’s parlance). You will need to have that anyway, as some record types have textual representations of their values.

I would probably implement the `Scanner` trait similar to the current `Scanner`. A subset of the methods in [Fundamental Scanning](https://docs.rs/domain/0.6.1/domain/master/scan/struct.Scanner.html#fundamental-scanning) as required methods and then the methods from [Complex Scanning](https://docs.rs/domain/0.6.1/domain/master/scan/struct.Scanner.html#complex-scanning) and more if necessary with default implementations. This will allow you to overwrite them if you want to, but you don’t really have to because, e.g. `scan_u16` will likely always be something like `u16::from_str(self.scan_word())`.

Error handling will be interesting, but I think we can use serde’s model here, too.

**Comment by not-my-profile** [comment]
Ah thanks, for elaborating. That makes sense. I have some time to hopefully implement this the next couple of days.

If we manage to resolve this issue, would you mind releasing a new version of the library, so that I can release my small educational name server on crates.io? If it doesn't fit your schedule, a prerelease e.g. `-beta` would also work fine.

**Comment by partim** [comment]
This should be possible.

Ideally, it should be possible to make the current `Scanner` implement the new `Scanner` trait, so that everything keeps mostly working. Most work will probably be to convert the current `Scan` impls for the various types, so maybe create an early draft PR so that we can discuss the design early on.

**Comment by not-my-profile** [comment]
Now that I am noticing all the Scanner intricacies, I am wondering if the format of the current zone file parser is based on a particular other implementation?

**Comment by partim** [comment]
The format is defined in [RFC 1035](https://tools.ietf.org/html/rfc1035). The definition could be more clear, but the parser follows is as well as possible.

**Comment by not-my-profile** [comment]
I've opened the PR #106 and am curious what you think of it.

**Comment by partim** [comment]
Not sure if you are still interested in this, but with the serde traits now implemented in #116, you can indeed parse record data from the command line using a dedicated deserializer.

A proof of concept is in this gist: https://gist.github.com/partim/10e71dbade4c1f1c5500d0343cf29dd4

As of now, `ZoneRecordData` requires the variant name (i.e., record type) to be capitalized like the variant name, but I am going to fix that.

**Comment by pspacek** [comment]
Just it case you wonder what else user might try to do while parsing RRsets, here is an attempt to make "maximally flexible" RRset parser in dnspython: https://dnspython.readthedocs.io/en/stable/zonefile.html

**Comment by partim** [comment]
A scanner that uses a list of strings as tokens was added in #142.
